### ISSUE TITLE
quotes and slash-quotes appearing in metadata strings

### ISSUE THREAD

**Issue by moodeaudio** [roderickvd/pleezer#46]
### Bug Description

I'm seeing single quotes and escaped single quotes in the metadata env vars set for the hook script.

```
pi@moode9:~ $ cat /var/local/www/deezmeta.txt 
'The Sound of Silence';'Pat Metheny';'What'\''s It All About';393;https://e-cdns-images.dzcdn.net/images/cover/5baa50987f37012e26b8a49e9380448f/500x500.jpg
```

The code block in the hook script that writes /var/local/www/deezmeta.txt is below
```
if [[ $EVENT == "track_changed" ]]; then
	METADATA=$TITLE";"$ARTIST";"$ALBUM_TITLE";"$DURATION";"https://e-cdns-images.dzcdn.net/images/cover/${ALBUM_COVER}/500x500.jpg"
	echo -e $METADATA > $DEEZMETA_FILE
	/var/www/util/send-fecmd.php "update_deezmeta,$METADATA"
fi
```

### Steps to Reproduce

Connect IOS Deezer app to moOde Deezer Connect

### Logs

I can provide log if needed :-)

### Operating System

PiOS

### pleezer version

v0.7.0 (and v0.6.2)

### Hardware specifications

Pi 5

### Deezer client specifications

iOS Deezer app

### Due Diligence

- [X] I am using the [latest version](https://github.com/roderickvd/pleezer/releases) of pleezer
- [X] I have searched the [issues](https://github.com/roderickvd/pleezer/issues) for similar reports
- [X] I have included a full verbose log from launch to the issue, not just an excerpt
- [X] I confirm that this is an issue with pleezer, not with my system configuration or other software
- [X] I confirm that this is not a security issue, which should be [reported](https://github.com/roderickvd/pleezer/security/policy) privately
- [X] I have read and understood the [Contributing guidelines](https://github.com/roderickvd/pleezer/blob/main/CONTRIBUTING.md)

**Comment by roderickvd** [comment]
This is actually not a bug but a feature: all environment variables are properly escaped for security. This means that you have to unescape them on your end.

Without shell escaping them, you would be vulnerable from command injection. For example, the following environment variable would exploit such a vulnerability: `TITLE=Bandits Gonna Trash Your Disk"; rm -rf / #`

You could do it in PHP or I believe the following would work (tested briefly):

```bash
if [ "$EVENT" = "track_changed" ]; then
    # Unescape all fields
    TITLE_UNESCAPED=$(eval "printf '%s' $TITLE")
    ARTIST_UNESCAPED=$(eval "printf '%s' $ARTIST")
    ALBUM_TITLE_UNESCAPED=$(eval "printf '%s' $ALBUM_TITLE")
    ALBUM_COVER_UNESCAPED=$(eval "printf '%s' $ALBUM_COVER")
    DURATION_UNESCAPED=$(eval "printf '%s' $DURATION")
    FORMAT_UNESCAPED=$(eval "printf '%s' $FORMAT")

    # Construct metadata with unescaped values
    METADATA="${TITLE_UNESCAPED};${ARTIST_UNESCAPED};${ALBUM_TITLE_UNESCAPED};${DURATION_UNESCAPED};https://e-cdns-images.dzcdn.net/images/cover/${ALBUM_COVER_UNESCAPED}/500x500.jpg"

    printf '%s\n' "$METADATA" > "$DEEZMETA_FILE"
    /var/www/util/send-fecmd.php "update_deezmeta,$METADATA"
fi
```

*Note:* this script does not take care of it, but you may want to wrap the fields in `deezmeta.txt` in quotes:
```
"The Sound of Silence";"Pat Metheny";"What's It All About";"393";"https://e-cdns-images.dzcdn.net/images/cover/5baa50987f37012e26b8a49e9380448f/500x500.jpg"
```

If you don't, and any field contains a `;` character, then your parsing will break:

```
The Sound of Silence;Pat Metheny; And Another Artist;What's It All About;393;https://e-cdns-images.dzcdn.net/images/cover/5baa50987f37012e26b8a49e9380448f/500x500.jpg
                                ^
                          breaks parsing
```

You would then have to unquote in the web frontend or whatever consumer of `deezmeta.txt`. In which case you would not need the unescaping in shell at all, because you already have it 😉

**Comment by moodeaudio** [comment]
Ok thanks, that explains it.

**Comment by roderickvd** [comment]
No worries, thanks for filing the issue nonetheless. For future reference what is your preferred approach? So I can think about how I can clarify it in the README.

**Comment by moodeaudio** [comment]
You mean the approach to implementing the unescaping?

I'll have to noodle it a bit and then get back to you.

**Comment by roderickvd** [comment]
Having slept over it, the script I proposed above is also not entirely safe. It is still exploitable with input like:
:warning: `TITLE=Bandits Gonna Trash Your Disk\"\;\ rm\ -rf\ /oh/no\ \\\"\#` ⚠️ 

I recommend that you unescape it in PHP with `stripcslashes($str)`.

**Comment by moodeaudio** [comment]
I think I would leave the metadata as-is (no shell escaping) and instead in the documentation note or warn that its the responsibility of the consuming application to sanitize the data. 

Here's some rationale (assumes the data is strictly used for display in UI)

1. Data thats shell escaped must first be unescaped by the consuming application which results in the original, potentially unsafe data.
```
php > echo stripcslashes('Bandits Gonna Trash Your Disk"; ls -l /oh/no #');
Bandits Gonna Trash Your Disk"; ls -l /oh/no #

php > echo stripcslashes("Bandits Gonna Trash Your Disk\"\;\ ls\ -l\ /oh/no\ \\\"\#");
Bandits Gonna Trash Your Disk"; ls -l /oh/no "#
```
This data then has to be further sanitized before using it which defeats the purpose of shell escaping it in the first place.

2. Using PHP stripcslashes() to process data thats shell escaped with inner quotes like this `'What'\''s It All About'` doesn't actually unescape the data correctly. For example:
```
# NOK with inner quotes
php > echo stripcslashes("'What'\''s It All About'");
'What'''s It All About'

# OK without inner quotes '\'')
php > echo stripcslashes("'What\'s It All About'");
'What's It All About'
``` 
I had to look back at how moode implements security for data that could be used in PHP exec() or shell_exec() or be processed in the WebUI by Javascript (JQuery). Whats done is to <em>reject the data</em> if it contains unwanted shell characters, directory traversal or certain shell commands. All PHP $_GET/$_POST data is checked by function chkValue() in inc/common.php.

Whats done for song file metadata that could potentially contain unsafe data is to HTML entity encode it in PHP before sending it to the WebUI for display for example `<script>unsafe code` becomes `&lt;script&gt;unsafe code` which is harmlessly displayed as text `<script>unsafe code` because JQuery in the Browser won't execute script tags formatted as `&lt;script&gt;`

Bottom line, at least for moode, is that I'd simply do PHP HTML entity encode on the as-is metadata before shipping it to the Browser. This mitigates harmful code being executed by Browser JS.

Also I prolly need to a different delimiter like \n instead of ;.

**Comment by roderickvd** [comment]
I would not be against changing it if it's better, and see your point for PHO and jQuery. But, how would you safeguard against such input in your shell script?

**Comment by moodeaudio** [comment]
From what I've read there are two basic ways to safeguard against executing unsafe code that's stored in shell variables.

1. Never use eval
2. Enclose the variable in double quotes to prevent shell characters from being interpreted
`echo "$VAR"` instead of `echo $VAR`

The hook script would have something like below
echo -e **"$METADATA"** > $DEEZMETA_FILE
/var/www/util/send-fecmd.php "update_deezmeta,$METADATA"

And then in `send-fecmd.php` I can apply htmlspecialchars($value, ENT_NOQUOTES) to the metadata. Its the same function used on song file metadata.

**Comment by roderickvd** [comment]
Yes, taking all this in, I think the safest setup would be to:
- stop escaping by pleezer
- change line 2 in your script to:
```bash
METADATA="${TITLE};${ARTIST};${ALBUM_TITLE};${DURATION};https://e-cdns-images.dzcdn.net/images/cover/${ALBUM_COVER}/500x500.jpg"
```

As for the delimiter, that's another issue. Changing the delimiter to `|` or something else just shifts the problem, and I've seen the weirdest track titles come by. To be robust I can see the following approaches (outside of pleezer):
1. escape the delimiter, like:
``` bash
# The escape_field function replaces ; with \;
escape_field() {
    echo "$1" | sed 's/;/\\;/g'
}
```
And then you'd have to unescape it.

3. URL-encode it by invoking some external interpreter:
```bash
url_encode() {
    # Python
    python3 -c "import urllib.parse; import sys; print(urllib.parse.quote(sys.argv[1], safe=''))" "$1"

    # PHP
    php -r "echo rawurlencode('$1');"

    # Perl
    perl -MURI::Escape -e 'print uri_escape($ARGV[0]);' "$1"
}

ENCODED=$(url_encode "Foo; Bar; Baz")
```
4. Encode then decode it like proper CSV.

**Comment by moodeaudio** [comment]
Ok cool :-)
The line 2 change is much cleaner.
For the delimiter I think I'm just going to use a rare character sequence like ~~~ or something like that until I get some time to redo things. The metadata  and its persistent file should prolly be JSON or YML format.

**Comment by roderickvd** [comment]
I've removed the shell escaping in 5587a974506482272032981302e073cc8b111d83...6df9404348c121c507998cd063cb38729191312b.

Latest version also supports podcasts and live radio. I intend to push a v0.8.0 release soon.

**Comment by moodeaudio** [comment]
Nice and wow about podcast and live radio. Thats really cool.

Thanks for info on the var change :-)

For live radio are the quality settings something that needs to be in a config screen?

**Comment by moodeaudio** [comment]
No prob, I'll add to the next release.
